### PR TITLE
MB-3913 Add updatePaymentServiceItemStatus event trigger

### PR DESCRIPTION
**Pull request by sandy-wright**
## Description

This PR adds a PaymentRequest.Update event trigger to the `updatePaymentServiceItemStatusHandler` in the GHC API. In addition, it fixes some error handling that this endpoint was missing previously (it wasn't catching stale eTags before). 

## Setup

First you'll need to run your local server and your local client:

```sh
make db_dev_e2e_populate && make server_run
make client_run
```

To test the GHC API using Postman, you'll also need to:

- Navigate to http://officelocal:3000. You should see a "Local Sign In" option - click this.
- Choose one of the default office user, TOO, or TIO emails and click "Login".
- Open up your developer tools and grab the following cookies (or enable Postman Interceptor and let that grab the cookies for you):
  - `office_session_token`
  - `masked_gorilla_csrf`
  - `_gorilla_csrf`
- Open Postman and click GHC API `updatePaymentServiceItemStatus` endpoint. Add the cookie values here by clicking the "Cookies" button on the right side (or, if using Postman Interceptor, make one dummy request using the endpoint and it will populate the cookie values for you).
- Now take the value for the `masked_gorilla_csrf` cookie and add a header called `X-CSRF-Token` and set it to the same value.
- Grab a Prime-available payment service item's ID and eTag.
  - You can do this by using the Prime API endpoint fetchMTOUpdates, which returns all Prime-available moves and their subobjects. Instructions for testing the Prime API here: https://github.com/transcom/prime_api_deliverable/wiki.
  - You may also use a combination of GHC endpoints to get this info (`listMTOs` then `listPaymentRequests`).
- Fill in the body with a new status, such as "APPROVED" or "DENIED", and submit the request!

If the request was successful, you should see an entry in the server log that says that a "Notification was created." You can also look directly in the database to verify that this did indeed happen.

## Code Review Verification Steps

* [x] Request review from a member of a different team.
* [x] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-3913) for this change

### CODE REVIEW

File: pkg/handlers/ghcapi/payment_service_items.go [hunk 1]
```diff
@@ -45,7 +47,8 @@ func (h UpdatePaymentServiceItemStatusHandler) Handle(params paymentServiceItemO
 
 	if err != nil {
 		logger.Error(fmt.Sprintf("Error finding payment service item for status update with ID: %s", params.PaymentServiceItemID), zap.Error(err))
-		return paymentServiceItemOp.NewUpdatePaymentServiceItemStatusNotFound()
+		payload := payloadForClientError("Unknown UUID(s)", "Unknown UUID(s) used to update a payment service item ", h.GetTraceID())
```
Comment (shimonacarvalho) [inline pkg/handlers/ghcapi/payment_service_items.go:50]: Awesome, fixed an empty payload!

File: pkg/handlers/ghcapi/payment_service_items.go [hunk 2]
```diff
@@ -78,7 +81,7 @@ func (h UpdatePaymentServiceItemStatusHandler) Handle(params paymentServiceItemO
 	// Using a switch to match error causes to appropriate return type in gen code
 	if err != nil {
 		switch err.(type) {
-		case services.PreconditionFailedError:
+		case query.StaleIdentifierError:
```
Comment (shimonacarvalho) [inline pkg/handlers/ghcapi/payment_service_items.go:84]: Good catch!

File: pkg/handlers/ghcapi/payment_service_items_test.go [hunk 3]
```diff
@@ -29,10 +31,10 @@ func (suite *HandlerSuite) TestUpdatePaymentServiceItemHandler() {
 
 	params := paymentServiceItemOp.UpdatePaymentServiceItemStatusParams{
 		HTTPRequest:          req,
-		IfMatch:              etag.GenerateEtag(paymentServiceItem.CreatedAt),
+		IfMatch:              etag.GenerateEtag(paymentServiceItem.UpdatedAt),
```
Comment (shimonacarvalho) [inline pkg/handlers/ghcapi/payment_service_items_test.go:34]: Another great catch.

File: pkg/handlers/ghcapi/payment_service_items_test.go [hunk 4]
```diff
@@ -110,15 +119,62 @@ func (suite *HandlerSuite) TestUpdatePaymentServiceItemHandler() {
...
-		suite.Equal(paymentServiceItem.ID.String(), okResponse.Payload.ID.String())
+		suite.Equal(deniedPaymentServiceItem.ID.String(), okResponse.Payload.ID.String())
 		suite.Equal(ghcmessages.PaymentServiceItemStatusAPPROVED, okResponse.Payload.Status)
 		suite.Nil(okResponse.Payload.RejectionReason)
 	})
 
+	suite.T().Run("Successful patch - Approval of Prime available paymentServiceItem", func(t *testing.T) {
+		availableMTO := testdatagen.MakeAvailableMove(suite.DB())
+		availablePaymentServiceItem := testdatagen.MakePaymentServiceItem(suite.DB(), testdatagen.Assertions{
+			PaymentRequest: testdatagen.MakePaymentRequest(suite.DB(), testdatagen.Assertions{
```
Comment (shimonacarvalho) [inline pkg/handlers/ghcapi/payment_service_items_test.go:137]: Technically you should be able to pass the availableMTO at the top level of the assertions I think, without having to make the nested objects yourself. But not all the testdatagen functions do that properly. 

You could give it a shot maybe? And just revert to this if it doesn't work. 
This is how it should work. And the `MakePaymentServiceItem` should know to create the nested item and pass on the assertion. 
```
testdatagen.MakePaymentServiceItem(suite.DB(), testdatagen.Assertions{
			Move: availableMTO,
}
```
Looking in the code for `MakePaymentServiceItem` it does have the calls to Make the nested items...
```
	paymentRequest := assertions.PaymentRequest
	if isZeroUUID(paymentRequest.ID) {
		paymentRequest = MakePaymentRequest(db, assertions)
	}

	mtoServiceItem := assertions.MTOServiceItem
	if isZeroUUID(mtoServiceItem.ID) {
		mtoServiceItem = MakeMTOServiceItem(db, assertions)
	}
```

File: pkg/handlers/ghcapi/payment_service_items_test.go [hunk 5]
```diff
@@ -110,15 +119,62 @@ func (suite *HandlerSuite) TestUpdatePaymentServiceItemHandler() {
...
+		suite.FatalNoError(err, "Error creating a new trace ID.")
+		handler.SetTraceID(traceID)
+
+		response := handler.Handle(params)
+		suite.IsType(&paymentServiceItemOp.UpdatePaymentServiceItemStatusOK{}, response)
+		okResponse := response.(*paymentServiceItemOp.UpdatePaymentServiceItemStatusOK)
+		suite.Equal(availablePaymentServiceItem.ID.String(), okResponse.Payload.ID.String())
+		suite.Equal(ghcmessages.PaymentServiceItemStatusAPPROVED, okResponse.Payload.Status)
+		suite.Nil(okResponse.Payload.RejectionReason)
+		suite.HasWebhookNotification(availablePaymentServiceItem.PaymentRequestID, traceID)
```
Comment (shimonacarvalho) [inline pkg/handlers/ghcapi/payment_service_items_test.go:178]: Ooh this is cool! I don't see the code for this in this pr - did you make it before?

File: pkg/handlers/ghcapi/payment_service_items.go [hunk 6]
```diff
@@ -88,7 +91,21 @@ func (h UpdatePaymentServiceItemStatusHandler) Handle(params paymentServiceItemO
 	if verrs != nil {
 		return paymentServiceItemOp.NewUpdatePaymentServiceItemStatusInternalServerError().WithPayload(&ghcmessages.Error{Message: handlers.FmtString(verrs.String())})
 	}
+
+	_, err = event.TriggerEvent(event.Event{
```
Comment (shimonacarvalho) [inline pkg/handlers/ghcapi/payment_service_items.go:95]: Looks perfect to me!

**Review by shimonacarvalho — APPROVED** [review]
Did leave one recommendation but that's optional, so approved!

**Comment by shimonacarvalho** [comment]
Oh one random note, maybe it would be good to figure out the normal way to test this in the UI - you know clicking an approve button somewhere, rather than working around it. If you login as TIO, you can choose a payment request that's already pending and see if there's something you can approve.

File: pkg/handlers/ghcapi/payment_service_items_test.go (hunk 4 above)
Comment (sandy-wright) [reply pkg/handlers/ghcapi/payment_service_items_test.go:137]: you're right, I changed it and it still worked perfectly! Nice, that makes things a lot simpler 😄

File: pkg/handlers/ghcapi/payment_service_items_test.go (hunk 5 above)
Comment (sandy-wright) [reply pkg/handlers/ghcapi/payment_service_items_test.go:178]: Yup, I made it for the `updatePaymentRequestStatus` endpoint - figured we're going to be doing this test a bunch more times, might as well make it a little simpler!

**Comment by sandy-wright** [comment]
> Oh one random note, maybe it would be good to figure out the normal way to test this in the UI - you know clicking an approve button somewhere, rather than working around it. If you login as TIO, you can choose a payment request that's already pending and see if there's something you can approve.

Yeah, that's a really good point. I think I'll try to set up a meeting with someone from TTV to go through how to test this in the interface because it's not totally apparent to me yet. Might be good for most of us to go to that, since we're going to be poking around these endpoints for a while!

**Pull request merged**